models/middlewares/dbChangesLogger.js

```diff
@@ -0,0 +1,74 @@
+import logger from "@config/logger";
+import { Changelog } from "@models/index";
+
+export default function dbChangesLoggerMiddleware(schema) {
+  let isUpdateOperation;
+
+  let username;
+  let collection;
+  let changesBefore;
+  let changesAfter;
+  let diff;
+
+  async function beforeUpdate(next) {
+    isUpdateOperation = this.op && this.op === "findOneAndUpdate";
+
+    username = isUpdateOperation 
+      ? this._conditions.username 
+      : this.username;
+    collection = isUpdateOperation 
+      ? this._collection.collectionName 
+      : this.constructor.collection.collectionName;
+    changesBefore = isUpdateOperation 
+      ? await this.model.findOne(this.getFilter()) 
+      : null;
+  
+    next();
+  }
+
+  async function afterUpdate(doc, next) {
+    changesAfter = await doc.constructor.findById(doc._id);
+
+    if (isUpdateOperation && changesBefore && changesAfter) {
+      diff = getChangesDiff(changesBefore._doc, changesAfter._doc);
+    } else {
+      diff = doc;
+    }
+
+    const change = {
+      username,
+      collectionName: collection,
+      collectionId: doc._id,
+      changesBefore,
+      changesAfter,
+      diff,
+    };
+
+    if (changesBefore !== changesAfter) {
+      await Changelog.create({ ...change });
+      logger.info(change);
+    } else {
+      logger.info(`No changes occured in: ${collection}`)
+    }
+
+    return next();
+  }
+  
+  schema.pre("save", beforeUpdate);
```

Comment:
I wonder if the automated tests fail due to a promise issue? In the docs for the newer version of mongoose it uses the example...

```js
schema.pre('save', async function() {
  await doStuff();
  await doMoreStuff();
});
```

https://mongoosejs.com/docs/middleware.html#pre

Reply:
It is just a suggestion, I am not sure why the tests file - it seems like a timeout each time and also random, which is why I think it might be a promise issue 🤔

Reply:
I also experienced similar `timeouts` scenarios with a playwright test I wrote recently in another project, It was my first/second time writing tests.

It turned out that the `timeouts` were not the reasons why the tests was failing, it was some `locators` and `assertions` that needed correcting. I have noticed similar pattern here with the tests too, permit me to try some things in the failing portion of the test.

### My Intentions

- Add a `wait` property to the browser `goto` function with a `domcontentloaded` value, this to override the default `load` value.

I would want to try this because, I think the `domcontentloaded` only fires when all html elements are fully loaded and parsed. Well put simple that the DOM has completed its work yea. 🤔

Reply:
We shouldn't need to add any waits/sleeps, these are considered bad practice. I think there is a race condition because different tests fail each time, which is why I think there is a promise not completing properly - as these tests passed correctly before 🤔

Reply:
I quite understand, I'm fairly new to testing. Intimidated! I just got the chance to turn my intimidation to strength just 2 weeks ago, when I wrote an `e2e` at https://github.com/open-sauced/app/pull/1551, hence my non-good practice thoughts, 

...but it is quite a thing to me for with the error in the stack trace:

```shell
3 failed
    [chromium] › search.spec.js:21:5 › Search works correctly ──────────────────────────────────────
    [chromium] › search.spec.js:35:5 › Search page has random results when no search term used ─────
    [chromium] › search.spec.js:46:5 › Search page shows random results after typing 1 characters ──
```

```shell
1 failed
    [chromium] › discover.spec.js:19:5 › Discover shows Profiles ───────────────────────────────────
```

I'm wondering, has there been any change made to these routes i.e. `/search` and `/discover`?